Sources/NIOSingletonsPosix/PosixSingletons.swift

```diff
@@ -0,0 +1,66 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_exported import NIOSingletonResources
+import NIOPosix
+import NIOCore
+
+extension NIOSingletons {
+    /// A globally shared, lazily initialized ``MultiThreadedEventLoopGroup``  that uses `epoll` as the selector mechanism.
+    ///
+    /// The number of threads is determined by ``NIOSingletons.suggestedMultiThreadedEventLoopGroupThreadCount``.
+    public static var multiThreadedPosixEventLoopGroup: MultiThreadedEventLoopGroup {
+        return globalMultiThreadedPosixEventLoopGroup
+    }
+
+    /// A globally shared, lazily initialized ``MultiThreadedEventLoopGroup`` with just one thread that uses `epoll` as the selector mechanism.
+    public static var singleThreadedPosixEventLoopGroup: MultiThreadedEventLoopGroup {
+        return globalSingleThreadedPosixEventLoopGroup
+    }
```

Comment:
I think we shouldn't offer both of these. If we're optimizing for non-expert users, we should make sensible choices for them: a user capable of answering the question "multi or single threaded" is also in a position to be able to find out how to arrange that behaviour.

Reply:
Also this makes the Rick Astley callback work a little better because we have two stanzas instead of 3.

Reply:
Right, I was thinking about this too. Yes, one option gives us a nicer API and we can likely recommend people to just go single-threaded.

What made choose to offer both is that it's entirely reasonable to start off single threaded and later decide or evaluate (not enough perf or so) if more threads would help. I liked that that experiment is then simply a single `s/single/multi/` and done.

If we only offered a single-threaded singleton then the step to trying a multi-threaded one would be quite steep because suddenly you need your own singletons (or pass around from main) and what not.

Of course we could just default to multi threaded (with a settable thread count if people really want to reduce it to 1) and not offer a single-threaded singleton ELG. I'd be fine with that.

The last option is to have just one singleton ELG that defaults to 1 thread but can be configured to have more than one thread (via env var / early-enough setting). The issue with that is that people will only test on the single-threaded one and suddenly a user changes some environment variable and now we're multi threaded which might break stuff as it's a config that has never run...

Reply:
> Also this makes the Rick Astley callback work a little better because we have two stanzas instead of 3.

We could also look up the next stanza ;p

Reply:
You know the rules, and so do I...

Reply:
I think it's extremely unclear to a library developer whether 1 or 1+ is appropriate. At the very least it isn't clear to me. What's appropriate for grpc-swift or AHC?

More importantly, the risk we have here is of multiplicative singletons. In your proposal we already have four:

1. NIOPosix.singleThreaded
2. NIOPosix.multiThreaded
3. NIOTS.singleThreaded
4. NIOTS.multiThreaded

Presuming a world where we have IOCP support we have two more, for 6.

I have no idea how a library developer is expected to pick which of these they should use. Even if we offered a "best available", we still have to offer "best available multithreaded" and "best available singlethreaded", and nothing prevents a developer from picking a different one. We rapidly trip into a world where we've created 2x core count threads for I/O and we haven't achieved a substantial simplification.

If we want NIO to have an easy mode, let's give it an easy mode. No choices. Users who aren't happy with that are still free to create ELGs, they are one line of code to create.

Reply:
Put another way, "consider how much you have in the ChannelPipeline" is not easy mode.

Reply:
Yes, it's not always easy. If you really think we should have just one then it has to be the multi-threaded.

I think it would be truly awful to give users one that's single-threaded by default (i.e. single threaded in all their tests) but can be turned into a multi-threaded ones by user config.

Given that a multi-threaded group doesn't cause harm it's gotta be the default if we only do one. I thought it might be nice to _additionally_ provide a single-threaded one in cases where you know that one thread will be enough. Happy for that affordance to go away especially given that you can do `singletonGroup.next()` to get  a single-threaded one.

Reply:
Single multi-threaded works for me.

Reply:
Fine by me.

Reply:
Done, implemented